File: android/app/src/main/kotlin/com/gps/tracker/BatteryEventHandler.kt
package com.gps.tracker

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.BatteryManager
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel

private const val CHANNEL = "battery_event_channel"

/**
 * Streams real-time battery level changes to Flutter via [EventChannel].
 *
 * Registers a [BroadcastReceiver] for [Intent.ACTION_BATTERY_CHANGED]
 * and forwards the percentage each time it fires.
 * No third-party battery packages are used.
 */
class BatteryEventHandler(
    private val context: Context,
    messenger: BinaryMessenger
) : EventChannel.StreamHandler {

    private val channel = EventChannel(messenger, CHANNEL)
    private var receiver: BroadcastReceiver? = null

    /** Registers this handler on the event channel. */
    fun register() {
        channel.setStreamHandler(this)
    }

    /** Unregisters the broadcast receiver to prevent leaks. */
    fun unregister() {
        receiver?.let { context.unregisterReceiver(it) }
        receiver = null
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
        receiver = object : BroadcastReceiver() {
            override fun onReceive(ctx: Context, intent: Intent) {
                val level = intent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1)
                val scale = intent.getIntExtra(BatteryManager.EXTRA_SCALE, -1)
                if (level >= 0 && scale > 0) {
                    events.success(level * 100 / scale)
                }
            }
        }
        context.registerReceiver(
            receiver,
            IntentFilter(Intent.ACTION_BATTERY_CHANGED)
        )
    }

    override fun onCancel(arguments: Any?) {
        unregister()
    }
}

File: ios/Runner/BatteryEventHandler.swift
import Flutter
import UIKit

private let kEventChannelName = "battery_event_channel"

/**
 * Streams real-time battery level changes to Flutter via `EventChannel`.
 *
 * Observes `UIDevice.batteryLevelDidChangeNotification` using
 * `UIDevice` battery APIs — no third-party packages are used.
 */
class BatteryEventHandler: NSObject, FlutterStreamHandler {

    private let eventChannel: FlutterEventChannel
    private var eventSink: FlutterEventSink?
    private var observer: NSObjectProtocol?

    init(messenger: FlutterBinaryMessenger) {
        eventChannel = FlutterEventChannel(name: kEventChannelName, binaryMessenger: messenger)
        super.init()
        eventChannel.setStreamHandler(self)
        UIDevice.current.isBatteryMonitoringEnabled = true
    }

    // MARK: - FlutterStreamHandler

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        observer = NotificationCenter.default.addObserver(
            forName: UIDevice.batteryLevelDidChangeNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            self?.sendBatteryLevel()
        }
        // Emit initial value immediately.
        sendBatteryLevel()
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        if let observer = observer {
            NotificationCenter.default.removeObserver(observer)
        }
        observer = nil
        eventSink = nil
        return nil
    }

    // MARK: - Private

    private func sendBatteryLevel() {
        let level = UIDevice.current.batteryLevel
        guard level >= 0 else { return }
        eventSink?(Int(level * 100))
    }
}
